### ISSUE TITLE
Validating classifier

### ISSUE THREAD

**Issue by arielah** [greenelab/wiki-nationality-estimate#4]
We still need a figure showing that our classifier works. Still not sure if we need a direct comparison to ethnicolr's performance.

**Comment by dhimmel** [comment]
From talking with @arielah, it seems like we'll have the following challenges directly comparing to ethnicolr:

- ethnicolr uses a different set of nationalities. And mapping between ours and theirs is not trivial

There are slight tweaks to the name preprocessing, but the classifier is pretty much the same. So we'd mostly be comparing the benefit of additional training examples. Therefore, I thought an alternative assessment we could make is to look at our model's performance as the number of training names increases. For example, a plot with "Number of training names" as the x-axis and "Testing performance" as the y-axis. Then we could note how much improvement in performance can be attributable to the increased sample size compared to ethnicolr's older dataset.

Let me look into the best metrics to use to assess predictions of a categorical variable (multiclass classification). @trang1618 / @cgreene / @mrgazzara any ideas?

**Comment by cgreene** [comment]
This works for me. In addition to amount of data I think there is probably also some effect of data quality. In a previous writeup from @arielah, I think she noted "Italian Wedding Soup" as a problematic example in the existing data that her approach avoided.

**Comment by trangdata** [comment]
In terms of multiclass classification metrics, I think weighted average F1 and balanced accuracy are pretty common. I have also used Cohen's Kappa coefficient before but that's a little harder to interpret.

**Comment by dhimmel** [comment]
> In terms of multiclass classification metrics, I think weighted average F1 is pretty common. I have also used Cohen's Kappa coefficient before but that's a little harder to interpret.

Good to know. I saw some options listed on [this sklearn page](https://scikit-learn.org/stable/modules/model_evaluation.html#multiclass-and-multilabel-classification) including a [log loss method](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.log_loss.html) that looks like it can take predicted probabilities.

My worry with F1 is that it required picking a single prediction for each observation. Would be great to have a method that assesses our predicted probabilities, since that's what we end up using in our analyses. I think Cohen's Kappa also has this issue.

**Comment by trangdata** [comment]
> My worry with F1 is that it required picking a single prediction for each observation. Would be great to have a method that assesses our predicted probabilities, since that's what we end up using in our analyses. I think Cohen's Kappa also has this issue.

Ah that's right I forgot we wanted to focus on the probabilities.

**Comment by dhimmel** [comment]
In https://github.com/greenelab/wiki-nationality-estimate/pull/5, @arielah computes log loss from the predicted probabilities as well as several metrics based on assigning each name a discrete class based on the max probability.

I'm wondering if there's a good way we could visualize the performance of the probabilities, since we end up using the probabilities and not a the max-probability class. What about an ROC curve for each nationality? An example is [given here](https://scikit-learn.org/stable/auto_examples/model_selection/plot_roc.html#plot-roc-curves-for-the-multilabel-problem). Basically we'd draw a separate ROC curve for each nationality. This does not assess the probabilities but instead the rank. Still it would provide a much richer and appropriate assessment.

We are tight on time, but I'd be happy to help here.

**Comment by trangdata** [comment]
Great idea! I have seen separate ROC curves like these before! Maybe during revision we can come back to log loss to compare performance of two methods on different training set size.

I'm happy to help as well!

**Comment by dhimmel** [comment]
We computed ROC curves in https://github.com/greenelab/wiki-nationality-estimate/pull/6 as well as AUROCCs. See AUROCCs for each category in https://github.com/greenelab/wiki-nationality-estimate/blob/master/models/LSTM_auroc.tsv. Really good performance, around 0.95.

@trang1618 did you want to plot ROC Curves? If so the data for the curves is in https://github.com/greenelab/wiki-nationality-estimate/blob/master/models/LSTM_roc_curves.tsv.

I'm a bit worried that all the curves will be on top of each other. It would get the point across that performance is high, but is not that useful for comparing curves. If we don't want a plot, we could just put the AUROCC values in a column in the nationality table.

**Comment by cgreene** [comment]
note for the future that since we're using probabilities it might be good to also plot calibration curves in the future:
https://scikit-learn.org/stable/auto_examples/calibration/plot_calibration_curve.html

**Comment by cgreene** [comment]
In summary, if I understand correctly, the figure in the paper will be a plot of AUROCs and a confusion matrix showing where errors occur?

**Comment by dhimmel** [comment]
> confusion matrix showing where errors occur?

I don't think we need confusion matrix at all. The ROCs / AUROCCs show a strong ability to predict all of the categories. The error rates in the confusion matrix are not indicative of our application, since we use the probabilities not max probability. Do you disagree?

> since we're using probabilities it might be good to also plot calibration curves in the future https://scikit-learn.org/stable/auto_examples/calibration/plot_calibration_curve.html

ooh yes those would be nice. The ROC curves show that the rankings are great, but we don't know whether the probabilities are properly tuned.

**Comment by cgreene** [comment]
@dhimmel : I guess I would still like to know where errors are made since this is multi-class. Maybe we can push this to think about on revision if we are invited to submit one. ROC + Calibration then?

**Comment by dhimmel** [comment]
> I guess I would still like to know where errors are made since this is multi-class.

We are plotting separate ROC curves for each class. Will touch base with Trang, but I don't think icorporating the ROC curves or AUROCCs will be too challenging.

**Comment by cgreene** [comment]
Yea - I guess I mean, when there is an error for "European" where does that error usually end up.

**Comment by trangdata** [comment]
I don't have prediction values to plot calibration curve, but ROC curves are here: https://github.com/greenelab/iscb-diversity/blob/76fcc6300c186e9e334e0e655fbacab5f271c019/figs/fig_3.png

**Comment by trangdata** [comment]
> Yea - I guess I mean, when there is an error for "European" where does that error usually end up.

I see what you mean, but since confusion matrix reflects a hard assignment for each observation instead of soft assignments as we used in the paper, I'm not sure if it's indicative of the method's error.

**Comment by dhimmel** [comment]
> ROC curves are here

Wow these are beautiful curves showing strong performance!

> I don't have prediction values to plot calibration curve

It probably would be a good idea to save our testing predictions to a TSV, for analyses like this. @arielah not sure if you'll get a chance to do that, but will save us time down the road.

Imagine a heatmap with true category as one axis and mean predicted probability as the other axis. Possibly normalized by category prevalence. That may best get at what Casey wants to assess.

**Comment by cgreene** [comment]
> I see what you mean, but since confusion matrix reflects a hard assignment for each observation instead of soft assignments as we used in the paper, I'm not sure if it's indicative of the method's error.

Is there some concept of "where does the probability for these examples end up" type of matrix? It would be like a soft confusion matrix.

**Comment by cgreene** [comment]
All of my comments of potentially new types of figures on this are for not the current revision. This one we should get the figure we most like inserted, the text written, and then we should ship it! 😁
